radixdlt-java/src/main/java/com/radixdlt/client/core/network/websocket/WebSocketClient.java

```diff
@@ -0,0 +1,140 @@
+import okhttp3.WebSocketListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the state for a single websocket.
+ */
+public class WebSocketClient implements PersistentChannel {
+	private static final Logger LOGGER = LoggerFactory.getLogger(WebSocketClient.class);
+	private final Object lock = new Object();
+	private final BehaviorSubject<WebSocketStatus> state = BehaviorSubject.createDefault(WebSocketStatus.DISCONNECTED);
+	private final Function<WebSocketListener, WebSocket> websocketFactory;
+	private final PublishSubject<String> messages = PublishSubject.create();
+	private WebSocket webSocket;
+
+	public WebSocketClient(Function<WebSocketListener, WebSocket> websocketFactory) {
+		this.websocketFactory = websocketFactory;
+		this.state
+			.filter(state -> state.equals(WebSocketStatus.FAILED))
+			.debounce(1, TimeUnit.MINUTES)
+			.subscribe(i -> {
+				synchronized (lock) {
+					if (this.state.getValue().equals(WebSocketStatus.FAILED)) {
+						this.state.onNext(WebSocketStatus.DISCONNECTED);
+					}
+				}
+			});
+	}
+
+	private WebSocket connectWebSocket() {
+		synchronized (lock) {
+			return this.websocketFactory.apply(new WebSocketListener() {
+				@Override
+				public void onOpen(WebSocket webSocket, Response response) {
+					synchronized (lock) {
+						WebSocketClient.this.state.onNext(WebSocketStatus.CONNECTED);
+					}
+				}
+
+				@Override
+				public void onMessage(WebSocket webSocket, String message) {
+					synchronized (lock) {
+						messages.onNext(message);
+					}
+				}
+
+				@Override
+				public void onClosing(WebSocket webSocket, int code, String reason) {
+					webSocket.close(1000, null);
+				}
+
+				@Override
+				public void onClosed(WebSocket webSocket, int code, String reason) {
+					synchronized (lock) {
+						WebSocketClient.this.state.onNext(WebSocketStatus.DISCONNECTED);
+						WebSocketClient.this.webSocket = null;
+					}
+				}
+
+				@Override
+				public void onFailure(WebSocket websocket, Throwable t, Response response) {
+					synchronized (lock) {
+						if (state.getValue().equals(WebSocketStatus.CLOSING)) {
+							WebSocketClient.this.state.onNext(WebSocketStatus.DISCONNECTED);
+							WebSocketClient.this.webSocket = null;
+							return;
+						}
+
+						WebSocketClient.this.state.onNext(WebSocketStatus.FAILED);
+						WebSocketClient.this.webSocket = null;
+					}
+				}
+			});
+		}
+	}
+
+	/**
+	 * Attempts to connect to this Radix node if not already connected
+	 */
+	public void connect() {
+		synchronized (lock) {
+			switch (state.getValue()) {
+				case DISCONNECTED:
+				case FAILED:
+					WebSocketClient.this.state.onNext(WebSocketStatus.CONNECTING);
+					this.webSocket = this.connectWebSocket();
+					return;
+				case CONNECTING:
+				case CONNECTED:
+				case CLOSING:
+				default:
+					return;
+			}
+		}
+	}
+
+	public Observable<String> getMessages() {
+		return messages;
+	}
+
+	public Observable<WebSocketStatus> getState() {
```

Comment:
This seems to be exposing too much of the internals, as callers could do something like `getState().onNext(someRandomState)` and break all the synchronization/mutexing that is happening in this class.

Better would be to expose just the functionality required, perhaps doOnNextState.

Reply:
They wouldn't be able to do that since this is exposing an `Observable` rather than a `Subject`